packages/fe-fpm-writer/src/page/index.ts

```diff
@@ -0,0 +1,100 @@
+import { join } from 'path';
+import { create as createStorage } from 'mem-fs';
+import { create, Editor } from 'mem-fs-editor';
+import { render } from 'ejs';
+
+import { enhanceData } from './defaults';
+import { CustomPage, InternalCustomPage, Ui5Route } from './types';
+import { getTemplateRoot } from './version';
+
+/**
+ * Add a new route to the provided route array, and update existing routes if necessary (e.g. if targets are defined as arrays for FCL).
+ *
+ * @param routes existing application routes (from the manifest)
+ * @param config configuration object
+ */
+function updateRoutes(routes: Ui5Route[], config: InternalCustomPage) {
+    const newRoute: Partial<Ui5Route> = {
+        name: `${config.entity}${config.name}`
+    };
+    if (config.navigation) {
+        const sourceRoute = routes.find((route) => route.name === config.navigation?.sourcePage);
+        newRoute.pattern = `${sourceRoute?.pattern.replace(':?query:', '')}/${config.navigation.navEntity}({${
+            config.navigation.navEntity
+        }Key}):?query:`;
+        if (sourceRoute?.target.constructor === Array) {
+            const pages = sourceRoute.target as string[];
+            // FCL only supports 3 columns, therefore, show the page in fullscreen if it is the 4th level of navigation
+            newRoute.target =
+                pages.length > 2 ? [(newRoute as Ui5Route).name] : [...pages, (newRoute as Ui5Route).name];
+        } else {
+            newRoute.target = newRoute.name;
+        }
+    } else {
+        newRoute.pattern = `${config.entity}({key}):?query:`;
+        newRoute.target = newRoute.name;
+    }
+    routes.push(newRoute as Ui5Route);
+}
+
+/**
+ * Validates the provided base path.
+ *
+ * @param {string} basePath - the base path
+ * @param {Editor} fs - the memfs editor instance
+ * @returns true if the path is valid, otherwise, throws and error
+ */
+export function validateBasePath(basePath: string, fs?: Editor): boolean {
```

Comment:
This validation makes sense for the other extension types as well

Reply:
3cbaa5b